drivers/src/pmu.rs

```diff
@@ -0,0 +1,155 @@
+                return Ok(ControlFlow::Break(()));
+            } else if !Self::validate_hardware_counter(info) {
+                return Err(SbiError::NotSupported);
+            }
+            self.hw_counters_info.push(Some(info));
+            self.valid_hw_counters_mask |= 1 << counter_index;
+        } else {
+            // Hardware counters can be sparse, and may not be implemented.
+            self.hw_counters_info.push(None);
+        }
+        Ok(ControlFlow::Continue(()))
+    }
+
+    /// Initializes the global PmuInfo structure with the information obtained from the platform SBI.
+    pub fn init() -> SbiResult<()> {
+        let mut pmu_info = PmuInfo::new();
+        let num_counters = pmu::get_num_counters()?;
+        for i in 0..num_counters.min(MAX_HARDWARE_COUNTERS as u64) {
+            if matches!(pmu_info.init_counter(i)?, ControlFlow::Break(())) {
+                break;
+            }
+        }
+        if !pmu_info.hw_counters_info.is_empty() {
+            PMU_INFO.call_once(|| pmu_info);
+            Ok(())
+        } else {
+            Err(SbiError::NotSupported)
+        }
+    }
+
+    /// Returns a reference to the global PmuInfo structure.
+    pub fn get() -> SbiResult<&'static PmuInfo> {
+        PMU_INFO.get().ok_or(SbiError::NotSupported)
+    }
+
+    /// Returns the counter index for the specified CSR
+    pub fn csr_to_counter_index(&self, csr: u64) -> SbiResult<u64> {
+        let counter_index = self
+            .hw_counters_info
+            .iter()
+            .position(|info| {
+                if let Some(info) = info {
+                    info.get_csr() == csr
+                } else {
+                    false
+                }
+            })
+            .ok_or(SbiError::InvalidParam)? as u64;
+        Ok(counter_index)
+    }
+
+    /// Returns the number of hardware counters supported by the platform
+    pub fn get_num_counters(&self) -> u64 {
+        self.hw_counters_info.len() as u64
+    }
+
+    /// Returns cached information for the hardware counter specified by counter_index.
+    pub fn get_counter_info(&self, counter_index: u64) -> SbiResult<PmuCounterInfo> {
+        let info = self
+            .hw_counters_info
+            .get(counter_index as usize)
+            .ok_or(SbiError::InvalidParam)?;
+        if let Some(info) = info {
+            Ok(*info)
+        } else {
+            Err(SbiError::InvalidParam)
+        }
+    }
+
+    /// Returns the CSR for the specified counter index
+    pub fn counter_index_to_csr(&self, counter_index: u64) -> SbiResult<u64> {
+        Ok(self.get_counter_info(counter_index)?.get_csr())
+    }
+
+    /// Validates the counter_index and counter_mask bit-mask for counter start and stop operations.
+    /// The counter_mask is relative to counter_index. If is_configuration_mask is false, the function
+    /// validates that any set bit in the counter_mask represents a valid hardware counter.
```

Comment:
Updated. BTW,  having gone over the implementation SBI once again (included below for reference), it seems like the best option for us is to simply check that no spurious bits are set, and let everything else through. Since the implementation errors out if the last set bit in the counter_mask is invalid, we don't really have to know which of the counters were actually started (for example). The workaround is to filter the passed in counter_mask, and mask bits for any counters that we know aren't present on the platform. That way, when the upper layer returns, we may get an already started/stopped error, but regardless, we only look at valid configured counters to update the state.

`int sbi_pmu_ctr_start(unsigned long cbase, unsigned long cmask,
                      unsigned long flags, uint64_t ival)
{
        int event_idx_type;
        uint32_t event_code;
        unsigned long ctr_mask = cmask << cbase;
        int ret = SBI_EINVAL;
        bool bUpdate = FALSE;

        if (sbi_fls(ctr_mask) >= total_ctrs)
                return ret;

        if (flags & SBI_PMU_START_FLAG_SET_INIT_VALUE)
                bUpdate = TRUE;

        for_each_set_bit_from(cbase, &ctr_mask, total_ctrs) {
                event_idx_type = pmu_ctr_validate(cbase, &event_code);
                if (event_idx_type < 0)
                        /* Continue the start operation for other counters */
                        continue;
                else if (event_idx_type == SBI_PMU_EVENT_TYPE_FW)
                        ret = pmu_ctr_start_fw(cbase, event_code, ival, bUpdate);
                else
                        ret = pmu_ctr_start_hw(cbase, ival, bUpdate);
        }

        return ret;
}`